### PR TITLE
chipsalliance/Surelog#2541: Add include file info model

### PR DESCRIPTION
**Pull request by hs-apotell**
Add include file info model

### CODE REVIEW

**Comment by hs-apotell** [comment]
@alaindargelas Could you please review the approach. Few follow up questions -

* IncludeFileInfo tracks both includes and macros. How to differentiate whether a particular entry belongs to include or macro?
* I am noticing an offset for generated column information. Values are offset-ed by -2. Any clue as to what might be the reason?

Or, instead of differentiating between include and macros, should the model itself have a type property so both include and macro become part of the model hierarchy. Macros will need to be exposed similarly at some point and this might be an opportunity to handle them in the same pass.

File: src/SourceCompile/SV3_1aPpTreeShapeListener.cpp [hunk 1]
```diff
@@ -556,25 +580,28 @@ void SV3_1aPpTreeShapeListener::enterMacroInstanceNoArgs(
...
       }
 
-      IncludeFileInfo info(macroInf->m_line, macroInf->m_file,
-                           m_pp->getSumLineCount() + 1, 0, 0, 0,
-                           IncludeFileInfo::PUSH);
-      m_pp->getSourceFile()->getIncludeFileInfo().push_back(info);
+      m_pp->getSourceFile()->getIncludeFileInfo().emplace_back(
+          macroInf->m_startLine, macroInf->m_file, m_pp->getSumLineCount() + 1,
+          macroInf->m_startColumn,
+          m_pp->getSumLineCount() + 1 +
```
Comment (alaindargelas) [inline src/SourceCompile/SV3_1aPpTreeShapeListener.cpp:594]: I can't do the mental math, but that might be where there is an offset created?

File: src/SourceCompile/SV3_1aPpTreeShapeListener.cpp [hunk 2]
```diff
@@ -621,23 +648,27 @@ void SV3_1aPpTreeShapeListener::enterMacroInstanceNoArgs(
...
       if (nbCRinMacroBody) {
-        IncludeFileInfo infop(line, fileId, m_pp->getSumLineCount() + 1, 0, 0,
-                              0, IncludeFileInfo::POP);
-        infop.m_indexOpening = openingIndex;
-        m_pp->getSourceFile()->getIncludeFileInfo().push_back(infop);
-        if (openingIndex >= 0)
+        m_pp->getSourceFile()->getIncludeFileInfo().emplace_back(
+            line, fileId, m_pp->getSumLineCount() + 1, startLineCol.first,
+            m_pp->getSumLineCount() + 1 +
+                (endLineCol.first - startLineCol.first),
```
Comment (alaindargelas) [inline src/SourceCompile/SV3_1aPpTreeShapeListener.cpp:662]: or here?

**Review by alaindargelas — COMMENTED** [review]
Don't make any change in the sv_vpi_user.h,
we have another file where we add the UHDM extensions:
templates/uhdm_vpi_user.h
Please move your changes from sv_vpi_user.h to uhdm_vpi_user.h

[include/sv_vpi_user.h](https://github.com/chipsalliance/UHDM/compare/5470f89838bb8f4ad80d6b939d3ae7e33b6230da...5b951a1f6dfc3204d265d22da780abde64a7fea9#diff-a27f14779b7fd9720238b314a93a03229e1589c9654c7870cf8c54edb7c6d8dc)

File: src/SourceCompile/SV3_1aPpTreeShapeListener.cpp (hunk 1 above)
Comment (hs-apotell) [reply src/SourceCompile/SV3_1aPpTreeShapeListener.cpp:594]: It takes the column information directly from value returned from `ParseUtils::getLineColumn` and `ParseUtils::getEndLineColumn`. Only line information is computed and that's coming out correct.

File: src/SourceCompile/SV3_1aPpTreeShapeListener.cpp (hunk 2 above)
Comment (hs-apotell) [reply src/SourceCompile/SV3_1aPpTreeShapeListener.cpp:662]: Same here.

File: src/SourceCompile/SV3_1aPpTreeShapeListener.cpp (hunk 1 above)
Comment (hs-apotell) [reply src/SourceCompile/SV3_1aPpTreeShapeListener.cpp:608]: This seems like a bug in ParseUtils logic. The two overloads for `getLineColumn` and `getEndLineColumn` aren't doing the same thing. Once I match the logic in both implementations, the result comes out of correct.

**Comment by hs-apotell** [comment]
> Please move your changes from sv_vpi_user.h to uhdm_vpi_user.h

Done.

Two questions -
1. `Preprocessor::m_includeFileInfo` collects both file includes and macros. How to differentiate between them? Should I add a enum to disambiguate?
2. In `UhdmWriter`, how do I get the _resolved_ file path? `IncludeFileInfo` only gives the text from the line in source file. How do I resolve this?

**Comment by alaindargelas** [comment]
1. Preprocessor::m_includeFileInfo collects both file includes and macros. How to differentiate between them? Should I add a enum to disambiguate?

Sure.

2. In UhdmWriter, how do I get the resolved file path? IncludeFileInfo only gives the text from the line in source file. How do I resolve this?

I need an example of what you get and what you expect to get.

**Comment by hs-apotell** [comment]
> I need an example of what you get and what you expect to get.

For an example, in PreprocFunc, what I get is "inc.h", what I need is the path to the file that this resolves to.

**Pull request merged**